File: wear/src/main/java/net/osdn/gokigen/joggingtimer/stopwatch/MyTimerTrigger.java
package net.osdn.gokigen.joggingtimer.stopwatch;

/**
 *
 *
 */
class MyTimerTrigger implements ITimerStopTrigger
{
    private final ITimeoutReceiver triggerReceiver;
    private final long duration;
    private final ITimerCounter timerCounter;
    private boolean forceStop = false;

    MyTimerTrigger(ITimeoutReceiver triggerReceiver, long duration, ITimerCounter timerCounter)
    {
        this.triggerReceiver = triggerReceiver;
        this.duration = duration;
        this.timerCounter = timerCounter;
    }

    void startTimer()
    {
        Thread thread = new Thread(() -> {
            if (timerCounter != null)
            {
                do {
                    try {
                        Thread.sleep(duration);
                    } catch (Exception e) {
                        // timeout! (But, do nothing!)
                    }
                    if (triggerReceiver != null)
                    {
                        triggerReceiver.timeout();
                    }
                } while ((timerCounter.isStarted())&&(!forceStop));
            }
            forceStop = false;
        });
        try
        {
            thread.start();
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }

    @Override
    public void forceStop()
    {
        forceStop = true;
    }

    public interface ITimeoutReceiver
    {
        void timeout();
    }
}

File: wear/src/main/java/net/osdn/gokigen/joggingtimer/stopwatch/timer/MyTimerTrigger.kt
package net.osdn.gokigen.joggingtimer.stopwatch.timer

import android.util.Log

/**
 *   タイマーカウンターの実装
 *
 */
class MyTimerTrigger(private val triggerReceiver: ITimeoutReceiver, private val duration: Long) : ITimerStopTrigger
{
    private var isStarted = false

    fun startTimer()
    {
        val thread = Thread {
            if (!isStarted)
            {
                isStarted = true
                Log.v(TAG, "MyTimerTrigger::startTimer() START")
                do
                {
                    try
                    {
                        Thread.sleep(duration)
                    }
                    catch (e: Exception)
                    {
                        // timeout! (But, do nothing!)
                    }
                    triggerReceiver.timeout()
                } while (isStarted)
            }
            else
            {
                Log.v(TAG, "MyTimerTrigger::startTimer() already started...")
            }
        }
        try
        {
            thread.start()
        }
        catch (e: Exception)
        {
            e.printStackTrace()
        }
    }

    // ITimerStopTrigger
    override fun forceStop()
    {
        isStarted = false
        Log.v(TAG, "MyTimerTrigger::forceStop()")
    }

    companion object
    {
        private val TAG = MyTimerTrigger::class.java.simpleName
    }

}
